src/hera/expr/_node.py

```diff
@@ -0,0 +1,224 @@
+    "not_in": "not in",
+    "starts_with": "startsWith",
+}
+
+BUILTINS = (
+    "all",
+    "any",
+    "count",
+    "filter",
+    "map",
+    "none",
+    "one",
+)
+
+
+class Node:
+    def __getattr__(self, name: str) -> "GetAttr":
+        return GetAttr(self, name)
+
+    def __getitem__(self, name: str) -> "GetItem":
+        return GetItem(self, name)
+
+    def __invert__(self):
+        return UnaryOp(self, "!")
+
+    def __neg__(self):
+        return UnaryOp(self, "-")
+
+    def __pos__(self):
+        return UnaryOp(self, "+")
+
+    def length(self):
+        return Callable("len", self)
+
+    def as_float(self):
+        return Callable("asFloat", self)
+
+    def as_int(self):
+        return Callable("asInt", self)
+
+    def check(self, truthy_value: 'Node', falsey_value: 'Node') -> 'Check':
+        return Check(self, truthy_value, falsey_value)
+
+    def get(self, name: str) -> "GetAttr":
+        return GetAttr(self, name)
+
+    def jsonpath(self, path: str):
+        return Callable("jsonpath", self, Constant(str(path)))
+
+    def string(self):
+        return Callable("string", self)
+
+    def to_json(self):
+        return Callable("toJson", self)
+
+
+for method, op in BINARY_OP_MAP.items():
+
+    def operator(op: str):
+        def func(self: Node, other: Node) -> "BinaryOp":
+            if not isinstance(other, Node):
+                other = Constant(other)
+            return BinaryOp(self, other, op)
+
+        return func
+
+    setattr(Node, method, operator(op))
+
+for builtin in BUILTINS:
+
+    def _builtin_func(builtin):
+        def func(self: Node, operation: Node) -> "Builtin":
+            return Builtin(builtin, self, operation)
+
+        return func
+
+    setattr(Node, builtin, _builtin_func(builtin))
+
+
+def _constant_repr(obj):
+    if obj is None:
+        return 'nil'
+    if obj is True:
+        return 'true'
+    if obj is False:
+        return 'false'
+    if isinstance(obj, range):
+        return f"{obj.start}..{obj.stop - 1}"
+    if isinstance(obj, list):
+        return f'[{", ".join(map(_constant_repr, obj))}]'
+    if isinstance(obj, dict):
+        key_value_pairs = [f"{_constant_repr(key)}: {_constant_repr(value)}" for key, value in obj.items()]
+        return f'{{{", ".join(key_value_pairs)}}}'
+    return repr(obj)
+
+
+class Constant(Node):
+    def __init__(self, value: Constants):
+        if isinstance(value, range):
+            if value.step != 1:
+                raise Exception("Only ranges with a step size of 1 are allowed")
+        self.value = value
+
+    def __repr__(self) -> str:
+        # we need a custom repr function in order to recursively translate
+        # True -> true, False -> false and None -> nil
+        return _constant_repr(self.value)
+
+
+class Identifier(Node):
+    def __init__(self, value: str = ""):
+        self.value = value
+
+    def __repr__(self) -> str:
+        return self.value
+
+
+class Parentheses(Node):
```

Comment:
I find `Constant` and `Identifier` conceptually sound in the sense that users are likely to have a 1:1 mapping between the _concept_ of a `Constant` and the _purpose_ of a `Constant`, which should facilitate usage. I have the same impression of `Identifier`. However, I am not so sure of `Parentheses`. I wonder if `Group` would be better conceptually? As a user if I want to group a statement I definitely _add_/_use_ parentheses. However, parentheses are an expression of a group rather than a group on its own. The group is a semantic element while the parentheses are a syntactic element. What do you think about the idea of using `Group`? If Argo uses the notion of `Parentheses` somewhere please point me to it! If that is the case then a 1:1 mapping with Argo is very reasonable and aligned with Hera post v4

Reply:
@flaviuvadan happy to rename this class. How would we want to expose it publicly? Currently it is exposed publicly via the class `P` in the hera.expr module.

Reply:
A global alias of `G`, perhaps?

Reply:
@flaviuvadan that might cause some confusion with the global context variable `g`. Are we okay with that?

Reply:
I originally thought the risk of confusion would not be too high but seeing [this line](https://github.com/argoproj-labs/hera-workflows/pull/340/files#diff-179cd0b6feab44a19c4a993d24c10669a890248331dfd25637bebdfbcf60ca9aR35) suggested, to me, that something like `g[...]` and `G[...]` _could_ both be syntactically correct in some instances but mean something very different, which might cause confusion and challenges debugging. I think we should leave `P` for `Parentheses` as is for now, and perhaps address this later if it _does_ turn out to be confusing in any way!